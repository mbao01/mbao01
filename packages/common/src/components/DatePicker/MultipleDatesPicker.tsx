"use client";

import { useState } from "react";
import { CalendarIcon } from "@radix-ui/react-icons";
import { format } from "date-fns";
import { cn } from "../../utilities";
import { Button } from "../Button";
import { Calendar } from "../Calendar";
import { Popover } from "../Popover";
import { type MultipleDatesPickerProps } from "./types";

export const MultipleDatesPicker = ({
  name,
  size,
  wide,
  label,
  outline,
  variant,
  disabled,
  defaultDates,
  getDatesValue = (dates) => dates?.map((date) => date.toUTCString()),
  getDatesLabel = (dates) => {
    if (!dates?.length) return undefined;
    if (dates.length < 3) return dates.map((date) => format(date, "LLL dd, y")).join(" and ");

    return `${dates.length} dates selected`;
  },
  ...props
}: MultipleDatesPickerProps) => {
  const [dates, setDates] = useState<Date[] | undefined>(defaultDates);

  const rangeLabel = getDatesLabel(dates);
  const datesValues = getDatesValue(dates);

  return (
    <Popover>
      {name &&
        datesValues?.map((date, index) => (
          <input
            key={date}
            hidden
            className="hidden"
            aria-hidden="true"
            name={`${name}.${index}`}
            defaultValue={date}
          />
        ))}
      <Popover.Trigger asChild>
        <Button
          name={name}
          size={size}
          wide={wide}
          outline={outline}
          variant={variant}
          disabled={disabled}
          className={cn("justify-start flex-nowrap", !dates?.length && "font-normal")}
        >
          <span className="text-left text-ellipsis overflow-hidden text-nowrap">
            {rangeLabel ?? label ?? "Pick one or more dates"}
          </span>
          <CalendarIcon className="ml-auto h-4 w-4 opacity-50" />
        </Button>
      </Popover.Trigger>
      <Popover.Content className="w-auto p-0">
        <Calendar
          autoFocus
          mode="multiple"
          selected={dates}
          onSelect={setDates}
          captionLayout="label"
          {...props}
        />
      </Popover.Content>
    </Popover>
  );
};
