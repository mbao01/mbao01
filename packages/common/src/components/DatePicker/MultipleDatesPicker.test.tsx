import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MultipleDatesPicker } from "./MultipleDatesPicker";

describe("MultipleDatesPicker", () => {
  it("shows the current month", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(<MultipleDatesPicker />);

    const button = screen.getByRole("button", {
      name: "Pick one or more dates",
    });

    await user.click(button);

    const date = new Date();
    const month = date.toLocaleString("default", { month: "long" });
    const year = date.getFullYear();

    expect(screen.getByText(`${month} ${year}`)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows default/preselected dates", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <MultipleDatesPicker
        defaultMonth={new Date("2022-03-01")}
        defaultDates={[new Date("2022-03-01"), new Date("2022-03-15")]}
      />
    );

    const button = screen.getByRole("button", {
      name: "Mar 01, 2022 and Mar 15, 2022",
    });

    await user.click(button);

    expect(screen.getByText("March 2022")).toBeInTheDocument();
    expect(screen.getByRole("gridcell", { name: "15" })).toHaveAttribute(
      "aria-selected",
      "true"
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows calendar with specific month", async () => {
    const user = userEvent.setup();

    const { asFragment } = render(
      <MultipleDatesPicker
        numberOfMonths={2}
        defaultMonth={new Date("2022-03-01")}
      />
    );

    const button = screen.getByRole("button", {
      name: "Pick one or more dates",
    });

    await user.click(button);

    expect(screen.getByText("March 2022")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a custom label", () => {
    const { asFragment } = render(
      <MultipleDatesPicker label="When will you be visiting?" />
    );

    expect(
      screen.getByRole("button", { name: "When will you be visiting?" })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("is disabled", () => {
    const { asFragment } = render(<MultipleDatesPicker disabled />);

    expect(
      screen.getByRole("button", { name: "Pick one or more dates" })
    ).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });
});
