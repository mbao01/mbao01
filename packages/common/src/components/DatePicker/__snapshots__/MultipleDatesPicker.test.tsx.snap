// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultipleDatesPicker > has a custom label 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r3»"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="closed"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      When will you be visiting?
    </span>
    <svg
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`MultipleDatesPicker > is disabled 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r4»"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="closed"
    disabled=""
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick one or more dates
    </span>
    <svg
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`MultipleDatesPicker > shows calendar with specific month 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r2»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick one or more dates
    </span>
    <svg
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`MultipleDatesPicker > shows default/preselected dates 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r1»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Mar 01, 2022 and Mar 15, 2022
    </span>
    <svg
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`MultipleDatesPicker > shows the current month 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-«r0»"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="btn min-h-fit h-10 rounded-md justify-start flex-nowrap font-normal"
    data-state="open"
    type="button"
  >
    <span
      class="text-left text-ellipsis overflow-hidden text-nowrap"
    >
      Pick one or more dates
    </span>
    <svg
      class="lucide lucide-calendar-days ml-auto h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 2v4"
      />
      <path
        d="M16 2v4"
      />
      <rect
        height="18"
        rx="2"
        width="18"
        x="3"
        y="4"
      />
      <path
        d="M3 10h18"
      />
      <path
        d="M8 14h.01"
      />
      <path
        d="M12 14h.01"
      />
      <path
        d="M16 14h.01"
      />
      <path
        d="M8 18h.01"
      />
      <path
        d="M12 18h.01"
      />
      <path
        d="M16 18h.01"
      />
    </svg>
  </button>
</DocumentFragment>
`;
